### PR TITLE
AB#103764 final changes for achievement date

### PR DESCRIPTION
**Pull request by dawid-pawulski**
### 🛠 Changes being made

- update unit tests
- add AchievementDate value object
- use modification tracker for milestone
- extend execute command to support errors

### 📸 Screenshots (optional)

image

### 🏎 Quality check

- [x] Have you performed local tests?
- [x] Are integration tests passing locally?
- [x] Have you added some unit tests?

### CODE REVIEW

File: src/HE.Investment.AHP.WWW/Models/AllocationClaim/MilestoneClaimModel.cs [hunk 1]
```diff
@@ -0,0 +1,37 @@
+using HE.Investments.AHP.Allocation.Contract.Claims.Enum;
+using HE.Investments.Common.Contract;
+
+namespace HE.Investment.AHP.WWW.Models.AllocationClaim;
+
+public class MilestoneClaimModel(
```
Comment (jjankowski87) [inline src/HE.Investment.AHP.WWW/Models/AllocationClaim/MilestoneClaimModel.cs:6]: Why do you need that model? It looks exactly the same as contract `MilestoneClaim`. If it's about setting `AchievementDate` later, then you can use `with` functionality of record

File: src/HE.Investment.AHP.WWW/Controllers/AllocationClaimController.cs [hunk 2]
```diff
@@ -181,17 +182,32 @@ private async Task<Phase> GetPhaseClaims(string allocationId, string phaseId, Ca
...
-            phaseClaims.MilestoneClaims.Single(x => x.Type == claimType));
+            milestoneClaimModel);
     }
 
     private async Task<IActionResult> ExecuteClaimCommand<TCommand>(
         TCommand command,
         string viewName,
-        CancellationToken cancellationToken)
+        CancellationToken cancellationToken,
+        DateDetails? achievementDate = null)
```
Comment (jjankowski87) [inline src/HE.Investment.AHP.WWW/Controllers/AllocationClaimController.cs:210]: Make execute command generic - pass some action argument

File: src/HE.Investments.AHP.Allocation.Domain/Claims/ValueObjects/DraftMilestoneClaim.cs [hunk 3]
```diff
@@ -24,16 +24,20 @@ public class DraftMilestoneClaim : MilestoneClaimBase
...
-        ValidatePreviousSubmissionDate(previousSubmissionDate, achievementDateTime);
-        ValidateProgrammeDates(programme, achievementDateTime!.Value);
+        ValidateDateIsNotFuture(achievementDate.Value, currentDate);
+        ValidatePreviousSubmissionDate(previousSubmissionDate, achievementDate.Value);
+        ValidateProgrammeDates(programme, achievementDate.Value!.Value);
+
+        ModificationTracker.Change(
+            ClaimDate.AchievementDate,
+            achievementDate,
+            () => ClaimDate.WithAchievementDate(achievementDate));
```
Comment (jjankowski87) [inline src/HE.Investments.AHP.Allocation.Domain/Claims/ValueObjects/DraftMilestoneClaim.cs:40]: This is value object so cannot be changed, create new object and return it. You don't need modification tracker.

File: src/HE.Investments.AHP.Allocation.Domain/Claims/ValueObjects/ClaimDate.cs [hunk 4]
```diff
@@ -5,7 +5,7 @@ namespace HE.Investments.AHP.Allocation.Domain.Claims.ValueObjects;
 
 public class ClaimDate : ValueObject
 {
-    public ClaimDate(DateTime forecastClaimDate, DateDetails? achievementDate = null, DateDetails? submissionDate = null)
+    public ClaimDate(DateTime forecastClaimDate, AchievementDate? achievementDate = null, DateDetails? submissionDate = null)
```
Comment (jjankowski87) [inline src/HE.Investments.AHP.Allocation.Domain/Claims/ValueObjects/ClaimDate.cs:8]: 3 dates and each of different type, so maybe it's a time to change SubmissionDate to DateTime (if cannot be modified) or VO if provided by Portal?

Comment (dawid-pawulski) [reply src/HE.Investments.AHP.Allocation.Domain/Claims/ValueObjects/ClaimDate.cs:8]: This will be covered in the next pr -> check answers and submit

**Review by Carq — APPROVED** [review]

**Review by jjankowski87 — APPROVED** [review]

**Pull request merged**